mayastor/tests/rebuild.rs

```diff
@@ -0,0 +1,1051 @@
+
+    // Remove the rebuild source.
+    let src_child = &get_share_uri(&ms2);
+    remove_child(nexus_hdl, src_child).await;
+    // Give a little time for the rebuild to fail.
+    std::thread::sleep(Duration::from_secs(1));
+    assert_eq!(get_num_rebuilds(nexus_hdl).await, 0);
+    // Nexus must be faulted because it doesn't have any healthy children.
+    assert_eq!(
+        get_nexus_state(nexus_hdl, NEXUS_UUID).await.unwrap(),
+        NexusState::NexusFaulted as i32
+    );
+}
+
+/// Test removing the destination of a rebuild.
+#[tokio::test]
+async fn rebuild_dst_removal() {
+    let test = start_infrastructure("rebuild_dst_removal").await;
+    let (mut ms1, _, ms3) = setup_test(&test, 1).await;
+    let nexus_hdl = &mut ms1;
+    let child = &get_share_uri(&ms3);
+
+    // Pause rebuild for added child.
+    add_child(nexus_hdl, child, true).await;
+    pause_rebuild(nexus_hdl, child).await;
+    assert!(wait_for_rebuild_state(
+        nexus_hdl,
+        child,
+        "paused",
+        Duration::from_secs(1),
+    )
+    .await
+    .unwrap());
+    assert_eq!(
+        get_nexus_state(nexus_hdl, NEXUS_UUID).await.unwrap(),
+        NexusState::NexusDegraded as i32
+    );
+
+    // Remove the child that is being rebuilt.
+    remove_child(nexus_hdl, child).await;
+    // Give a little time for the rebuild to fail.
+    std::thread::sleep(Duration::from_secs(1));
+    assert_eq!(get_num_rebuilds(nexus_hdl).await, 0);
+    // Nexus must be online because it has a single healthy child.
+    assert_eq!(
+        get_nexus_state(nexus_hdl, NEXUS_UUID).await.unwrap(),
+        NexusState::NexusOnline as i32
+    );
+}
+
+/// Test faulting the source of a rebuild.
+#[tokio::test]
+async fn rebuild_fault_src() {
+    let test = start_infrastructure("rebuild_fault_src").await;
+    let (mut ms1, mut ms2, ms3) = setup_test(&test, 1).await;
+    let nexus_hdl = &mut ms1;
+    let child = &get_share_uri(&ms3);
+
+    // Check a rebuild is started for the added child.
+    add_child(nexus_hdl, child, true).await;
+    assert!(wait_for_rebuild_state(
+        nexus_hdl,
+        child,
+        "running",
+        Duration::from_millis(500),
+    )
+    .await
+    .unwrap());
+
+    // Fault the rebuild source by unsharing the bdev.
+    bdev_unshare(&mut ms2).await;
+
+    // The rebuild failed so the destination should be faulted.
+    assert!(
+        wait_for_child_state(
+            nexus_hdl,
+            child,
+            ChildState::ChildFaulted,
+            Duration::from_millis(500),
+        )
+        .await
+    );
+    assert_eq!(get_num_rebuilds(nexus_hdl).await, 0);
+}
+
+/// Test faulting the destination of a rebuild.
+#[tokio::test]
+async fn rebuild_fault_dst() {
+    let test = start_infrastructure("rebuild_fault_dst").await;
+    let (mut ms1, _, mut ms3) = setup_test(&test, 1).await;
+    let nexus_hdl = &mut ms1;
+    let child = &get_share_uri(&ms3);
+
+    // Check a rebuild is started for the added child.
```

Comment:
this pattern seems to repeat quite a bit, maybe could be made into a single fn

Reply:
Yeah I thought about that but there were some tests where additional behaviour was required between adding a child and rebuilding (i.e. pausing) which is why I didn't separate it out.

I'll leave it like this for now and revisit it if needed.